ziti/src/main/kotlin/org/openziti/identity/Enroller.kt

```diff
@@ -16,38 +16,69 @@
 
 package org.openziti.identity
 
-import com.google.gson.Gson
-import com.google.gson.JsonObject
 import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.asExecutor
 import kotlinx.coroutines.runBlocking
 import org.bouncycastle.asn1.x500.X500Name
+import org.bouncycastle.pkcs.PKCS10CertificationRequest
 import org.bouncycastle.pkcs.jcajce.JcaPKCS10CertificationRequestBuilder
-import org.bouncycastle.util.io.pem.PemObject
-import org.bouncycastle.util.io.pem.PemWriter
+import org.openziti.Enrollment
+import org.openziti.IdentityConfig
+import org.openziti.edge.ApiClient
+import org.openziti.edge.ApiException
+import org.openziti.edge.api.AuthenticationApi
+import org.openziti.edge.api.ControllersApi
+import org.openziti.edge.api.InformationalApi
+import org.openziti.edge.model.ApiErrorEnvelope
+import org.openziti.edge.model.EnrollmentCertsEnvelope
 import org.openziti.util.*
-import java.io.ByteArrayOutputStream
-import java.io.OutputStreamWriter
-import java.net.URL
-import java.net.URLEncoder
+import java.io.File
+import java.net.URI
+import java.net.http.HttpClient
+import java.net.http.HttpRequest
+import java.net.http.HttpResponse
+import java.security.KeyPair
 import java.security.KeyPairGenerator
-import java.security.KeyStore
+import java.security.PrivateKey
 import java.security.SecureRandom
+import java.security.cert.CertificateFactory
 import java.security.cert.X509Certificate
 import java.security.spec.ECGenParameterSpec
-import javax.net.ssl.*
-import kotlin.text.Charsets.UTF_8
-
-class Enroller(
-    val enrollmentURL: URL,
-    val method: Method,
-    val name: String,
-    val caCerts: Collection<X509Certificate>
-) : Logged by ZitiLog("ziti-enroler") {
-
-    enum class Method {
-        ott,
-        ottca,
-        ca
+import java.util.*
+import java.util.concurrent.TimeUnit
+import java.util.function.Consumer
+import javax.net.ssl.SSLContext
+
+internal class Enroller(
+    private val jwt: ZitiJWT
+) : Enrollment, Logged by ZitiLog("ziti-enroll") {
+
+    private val http: HttpClient.Builder = ApiClient.createDefaultHttpClientBuilder().apply {
+        executor(Dispatchers.IO.asExecutor())
+        sslContext(SSLContext.getInstance("TLSv1.2").apply {
+            init(null, arrayOf(KeyTrustManager(jwt.serverKey)), SecureRandom())
+        })
+    }
+
+    private val api: ApiClient =
+        ApiClient(http, ApiClient.createDefaultObjectMapper(), jwt.controller.toString())
+
+    val name = jwt.name
+    val token = jwt.token
+
+    private val caCerts: Collection<X509Certificate>
+
+    init {
+        val wellKnownCertsReq = HttpRequest.newBuilder(jwt.controller.resolve("/.well-known/est/cacerts")).build()
```

Comment:
did you test this when the controller uses a public CA that's already in the OS trust store? the ziti cli handles it mostly, but i've been finding that particular situation isn't one we end up testing much, we almost always stay with the private CA we generate

Reply:
it is no needed here. server cert is verified against JWT